nerfstudio/model_components/renderers.py

```diff
@@ -130,59 +134,60 @@ def get_background_color(
+        """Blends the background color into the image if image is RGBA.
+        Otherwise no blending is performed (we assume opacity of 1).
 
         Args:
-            rgb: RGB per pixel.
+            image: RGB/RGBA per pixel.
             opacity: Alpha opacity per pixel.
             background_color: Background color.
 
         Returns:
             Blended RGB.
         """
-        if len(opacity.shape) < len(rgb.shape):
-            opacity = opacity.unsqueeze(-1)
+        if image.size(-1) < 4:
+            return image
+
+        if background_color is None:
+            background_color_name = self.background_color
+            if background_color_name == "last_sample" or background_color_name == "random":
+                background_color_name = "black"
+            background_color = self.get_background_color(
+                rgb=image.unsqueeze(-1), background_color=background_color_name
+            )
+        rgb, opacity = image[..., :3], image[..., 3:]
         blended_rgb = rgb * opacity + background_color.to(rgb.device) * (1 - opacity)
         return blended_rgb
 
-    def blend_background_for_metric_computation(
+    def blend_background_for_loss_computation(
         self,
+        pred_image: Tensor,
+        pred_accumulation: Tensor,
         gt_image: Tensor,
-        predicted_rgb: Tensor,
-        predicted_opacity: Tensor,
-        background_color: Float[Tensor, "3"] = colors.BLACK,
     ) -> Tuple[Tensor, Tensor]:
         """Blends a background color into the ground truth and predicted image for
-        metric computation if the current background color is "random".
+        loss computation.
 
         Args:
             gt_image: The ground truth image.
-            predicted_rgb: The predicted RGB values.
-            predicted_opacity: The predicted opacity/ accumulation.
-            background_color: The background color to use for blending. This should be
-                a single color to use for all pixels.
-
+            pred_image: The predicted RGB values (without background blending).
+            pred_accumulation: The predicted opacity/ accumulation.
         Returns:
             A tuple of the predicted and ground truth RGB values.
         """
-        if gt_image.shape[-1] == 4 and self.background_color == "random":
-            predicted_rgb = self.blend_background(
-                rgb=predicted_rgb, opacity=predicted_opacity, background_color=background_color
-            )
-            gt_rgb = self.blend_background(
-                rgb=gt_image[..., :3], opacity=gt_image[..., 3], background_color=background_color
-            )
-        else:
-            gt_rgb = gt_image[..., :3]
 
-        return predicted_rgb, gt_rgb
+        background_color = self.get_background_color(
+            rgb=pred_image.unsqueeze(-1), background_color=self.background_color
+        ).to(pred_image.device)
+        if self.background_color == "random":
```

Comment:
Looks like this if statement is still there from the last change

Reply:
Now I think it is correct, no? The predicted RGB is only "blended" if self.background_color == random. The GT is always blended (as it can be transparent here).

Reply:
Ah, I see what you mean. So this is resolving the double blending issue by blending "random" here and everything else as done before?

Reply:
In this case, I think it is fine like this and no need to do anything for https://github.com/nepfaff/nerfstudio/pull/1#discussion_r1265175381